indexing-service/src/main/java/io/druid/indexing/common/actions/SegmentTransactionalInsertAction.java

```diff
@@ -99,14 +99,24 @@ public DataSourceMetadata getEndMetadata()
   @Override
   public SegmentPublishResult perform(Task task, TaskActionToolbox toolbox) throws IOException
   {
-    toolbox.verifyTaskLocks(task, segments);
+    for (DataSegment segment : segments) {
+      if (!toolbox.getTaskLockbox().upgrade(task, segment.getInterval()).isOk()) {
+        return SegmentPublishResult.fail();
+      }
+    }
+
+    TaskActionPreconditions.checkTaskLocks(task, toolbox.getTaskLockbox(), segments);
```

Comment:
Nice catch! Yes it should be. But now, it looks better to check locks first and then upgrade them. ```checkTaskLocksUpgraded()``` is not used anymore and I removed it.